compiler/src/codegen/compcore.re

```diff
@@ -1124,56 +1136,136 @@ let compile_array_op = (wasm_mod, env, arr_imm, op) => {
     // incref/decref pair). Since it won't live in a local, it wouldn't be
     // cleaned up automatically anyway.
     compile_imm(~skip_incref=true, wasm_mod, env, arr_imm);
-  switch (op) {
-  | MArrayGet(idx_imm) =>
-    // ASSUMPTION: idx is a basic (non-heap) int
-    let idx = compile_imm(wasm_mod, env, idx_imm);
-    let set_idx = () => set_swap(1, untag_number(wasm_mod, idx));
-    let get_idx = () => get_swap(1);
-    let set_arr = () => set_swap(2, get_arr_value());
-    let get_arr = () => get_swap(2);
-    /* Check that the index is in bounds */
+  let resolve_idx = () => {
+    // PRECONDITION: idx is in swap slot 1
+    // PRECONDITION: arr is in swap slot 2
     Expression.Block.make(
       wasm_mod,
-      gensym_label("MArrayGet"),
+      gensym_label("resolve_idx"),
       [
-        set_idx(),
-        set_arr(),
-        /*
-         Check index not out of bounds (negative end)
-         */
-        error_if_true(
+        /* Maximum array length in 32-bit mode is less than 2^30, so any heap-allocated int is out of bounds. */
+        /* Check that the index is a simple int. */
+        Expression.If.make(
+          wasm_mod,
+          Expression.Unary.make(
+            wasm_mod,
+            Op.eq_z_int32,
+            Expression.Binary.make(
+              wasm_mod,
+              Op.and_int32,
+              get_swap(1),
+              Expression.Const.make(wasm_mod, const_int32(1)),
+            ),
+          ),
+          Expression.Block.make(
+            wasm_mod,
+            gensym_label("IndexNotSimpleInteger"),
```

Comment:
I feel like this is a symptom that `XX.toNumber` should reduce its argument to the smallest containing integer, rather than be a no-op...

Reply:
I think how it works now is correct (since the flexibility is the point of Number), but we should definitely have a utility that does let you round/truncate. It gets kinda hairy for floats, since we support bigints. My basic math tells me that the worst case conversion of a float64 to a bigint would be a bigint with 16 limbs, so maybe not actually so bad.

That said though, I could see us converting whole numbers to ints. `3.0` becoming `3`, that sort of thing.

But yeah, food for thought.